pipelines/curate_identity.py

```
@@ -0,0 +1,547 @@
+        if result.action == "retained":
+            source = result.mapping["source"]
+            for id_mapping in result.mapping["id_mappings"]:
+                output_id = id_mapping["output_id"]
+                owner_path = id_mapping["owner_path"]
+                if output_id in seen:
+                    first, first_owner = seen[output_id]
+                    raise CanonicalIdCollision(
+                        f"canonical ID collision {output_id!r}: "
+                        f"{first['path']}:{first['line']}{first_owner} and "
+                        f"{source['path']}:{source['line']}{owner_path}"
+                    )
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)